src/coro/coro.ml

```diff
@@ -318,7 +328,10 @@ let fun_to_coro ctx coro_type =
 	]) basic.tvoid null_pos in
 
 	let tf_args = args @ [ (vcompletion,None) ] in
-	let tf_type = basic.tcoro.continuation_result in
+	(* I'm not sure what this should be, but let's stick to the narrowest one for now.
+	   Cpp dies if I try to use coro_class.outside.cls_t here, which might be something
+	   to investigate independently. *)
+	let tf_type = basic.tcoro.continuation_result coro_class.outside.result_type in
```

Comment:
I think the return type being `ContinuationResult` is fine, Ideally the generated continuation class shouldn't need to leak out of that function.

I had a look at what cpp error happens, seems to be another case of the class not being forward declared. `cppReferences` will need to be updated if we do want the specific type returned.

Reply:
I agree, let's leak as little as possible. After all, that's why I've create ContinuationResult (which should probably actually be CoroutineResult?) in the first place.

Reply:
By the way, I always get very confused when reading Kotlin docs because most of the time when they talk about "coroutine", they actually mean the entire thing that we currently create via `Coroutine.run`. The functions that we use `@:coroutine` on are just suspend functions in Kotlin, but this whole terminology isn't used consistently in Kotlin world either.

I just mention this because a name like `CoroutineResult` could be misunderstood to mean the result of a `Coroutine.run`. So I don't know what to call anything really.

Reply:
I also find myself getting tripped up by the name of things quite often. Especially continuations vs completions and the fact that one of the core interfaces is `IContinuation. I also do not have any better ideas for names at this point.